Add hierarchical discriminator multiplicity support

kugan-nv opened the pull request:
Added support for mutiplicity and copy_id for gcc gcov. This should matches the GCC
implementation. I am posting patches to GCC to match this.

This patch also adds support for two-pass aggregation when using hierarchical
discriminator encoding (gcov_version=3). The two-pass approach correctly
handles duplicate samples from code duplication (loop unrolling, etc.).

  * GetBaseDiscriminator() - Extract bits 0-7
  * GetMultiplicity() - Extract bits 8-14 (duplication factor)
  * GetCopyID() - Extract bits 15-25 (unused but reserved)

With the patch we now:
1. Extract multiplicity and copy_id from discriminator
2. Multiply sample count by multiplicity
4. Add samples  using copy_id.

kugan-nv committed:
Add hierarchical discriminator multiplicity support

Added support for mutiplicity and copy_id for gcc gcov. This should matches the GCC
implementation. I am posting patches to GCC to match this.

This patch also adds support for two-pass aggregation when using hierarchical
discriminator encoding (gcov_version=3). The two-pass approach correctly
handles duplicate samples from code duplication (loop unrolling, etc.).

  * GetBaseDiscriminator() - Extract bits 0-7
  * GetMultiplicity() - Extract bits 8-14 (duplication factor)
  * GetCopyID() - Extract bits 15-25 (unused but reserved)

With the patch we now:
1. Extract multiplicity and copy_id from discriminator
2. Multiply sample count by multiplicity
4. Add samples  using copy_id.

This was already supported for LLVM.

kugan-nv committed:
Add hierarchical discriminator multiplicity support

Added support for mutiplicity and copy_id for gcc gcov. This should matches the GCC
implementation. I am posting patches to GCC to match this.

This patch also adds support for two-pass aggregation when using hierarchical
discriminator encoding (gcov_version=3). The two-pass approach correctly
handles duplicate samples from code duplication (loop unrolling, etc.).

  * GetBaseDiscriminator() - Extract bits 0-7
  * GetMultiplicity() - Extract bits 8-14 (duplication factor)
  * GetCopyID() - Extract bits 15-25 (unused but reserved)

With the patch we now:
1. Extract multiplicity and copy_id from discriminator
2. Multiply sample count by multiplicity
4. Add samples  using copy_id.

This was already supported for LLVM.

erozenfeld commented on source_info.h:
```diff
@@ -79,6 +88,11 @@ struct SourceInfo {
   }
+  
+  // Return the full 64-bit offset for 32-bit hierarchical discriminators.
+  static constexpr uint64_t GenerateFullOffset(uint64_t Offset) {
```
This is not called from anywhere , is it?

kugan-nv replied:
Removed this

erozenfeld reviewed (approved):
LGTM with one question

kugan-nv commented:
Thanks for the review @erozenfeld.

I have also added  support for aggregating samples with same with same [line + base discriminator] as a second pass as GCC doesn't aggregate.  This is only done for create_gcov (for gcc) and supported with a flag. This is also enabled only for gcov version 3.

GCC patches are now comitted.

kugan-nv commented:
@erozenfeld: Ping ?

snehasish commented:
Is it possible to add some basic tests for this?

kugan-nv commented:
Updated the patch with a test. Added a binary and perf.data generated on x86 machine with the gcc 16. The test checks for the hierarchical discriminator processing for GCC.

snehasish commented on testdata/hierarchical_discriminator_test.c:
```diff
@@ -0,0 +1,42 @@
+
+  for (i = 0; i < 32; i++) {
+      for (j = 0; j < 4; j++)
+	sum += a[i + j*4] * i;
```
indentation here is misleading

snehasish commented on testdata/hierarchical_discriminator_test.c:
```diff
@@ -0,0 +1,42 @@
+
+float vx[2048];
+float vy[2048];
+float vz[2048];
```
I think if we use this without initialization then the compiler may optimize it away.

snehasish commented on symbol_map.h:
```diff
@@ -36,6 +36,11 @@
 // Whether to use discriminator encoding.
 ABSL_DECLARE_FLAG(bool, use_discriminator_encoding);
 
+// Whether to use two-pass aggregation for copy_id:
```
note that this is gcov only?

snehasish commented on profile_writer.cc:
```diff
@@ -257,6 +257,11 @@ bool AutoFDOProfileWriter::WriteToFile(const std::string &output_filename) {
+  // Pass 2: Collapse copy_ids before writing (SUM aggregation)
+  if (absl::GetFlag(FLAGS_use_two_pass_aggregation)) {
+    symbol_map_->CollapseCopyIDs();
+  }
```
I think this is the reason why we have change the profile writer constructors to non-const. Is there a way to preserve the original const contract by refactoring the implementation?

snehasish reviewed (approved):
lgtm with some minor comments.

kugan-nv commented:
Ping? Thanks

snehasish commented:
Hi @kugan-nv,
The CI checks currently appear stuck. I tried re-running all jobs, but it didn't seem to trigger them. Not sure whats happening. Could you try either a close and reopen of the PR, or pushing an empty commit to see if that kicks off the workflow runners?
